app/uk/gov/hmrc/gform/auth/AuthService.scala

```diff
@@ -78,32 +78,88 @@ class AuthService(
   private val notAuthorized: AuthResult = AuthBlocked("You are not authorized to access this service")
   private val decoder = Base64.getDecoder
 
-  private def performAWSALBAuth()(implicit hc: HeaderCarrier): AuthResult = {
+  private def performAWSALBAuth(assumedIdentity: Option[String])(implicit hc: HeaderCarrier): AuthResult = {
+    Logger.info("ALB-AUTH: Start authorization...")
+
     val encodedJWT: Option[String] = hc.otherHeaders.collectFirst {
-      case (header, value) if header === "x-amzn-oidc-data" => value
+      case (header, value) if header === "X-Amzn-Oidc-Data" => value
     }
 
+    Logger.info(s"ALB-AUTH: JWT -> [${encodedJWT.getOrElse("No ALB JWT")}]")
+
     encodedJWT.fold(notAuthorized) { jwt =>
       jwt.split("\\.") match {
         case Array(header, payload, signature) =>
           val payloadJson = new String(decoder.decode(payload))
           Try(Json.parse(payloadJson)) match {
             case Success(json) =>
-              json \ "username" match {
-                case JsDefined(JsString(username)) => AuthSuccessful(AWSALBRetrievals(username))
-                case _                             => AuthBlocked("Username does not exist in JWT")
+              Json.fromJson[JwtPayload](json) match {
+                case JsSuccess(jwtPayload, _) => {
+                  Logger.info(s"ALB-AUTH: Authorizing with following credentials : [${jwtPayload.toString}]")
+                  AuthSuccessful(awsAlbAuthenticatedRetrieval(jwtPayload, assumedIdentity))
+                }
+                case JsError(_) => AuthBlocked("Not authorized")
               }
             case Failure(_) =>
-              Logger.error(s"Corrupt JWT received from AWS ALB: payload is not a json: $payloadJson")
+              Logger.error(s"ALB-AUTH : Corrupt JWT received from AWS ALB: payload is not a json: $payloadJson")
               notAuthorized
           }
         case _ =>
-          Logger.error(s"Corrupt JWT received from AWS ALB: [$jwt]")
+          Logger.error(s"ALB-AUTH : Corrupt JWT received from AWS ALB: [$jwt]")
           notAuthorized
       }
     }
   }
 
+  private def awsAlbAuthenticatedRetrieval(
+    jwtPayload: JwtPayload,
+    assumedIdentity: Option[String]
+  ): AuthenticatedRetrievals =
+    if (jwtPayload.iss == appConfig.albAdminIssuerUrl) {
+
+      val identity = assumedIdentity match {
+        case Some(user) => user
+        case None       => jwtPayload.username
+      }
+
+      Logger.info(s"ALB-AUTH: Admin will assume the following user identity : [$identity]")
+
+      AuthenticatedRetrievals(
```

Comment:
`AuthenticatedRetrievals` in this `if` branch  differs only by two parameters from the one in `else` branch.

Creation of `AuthenticatedRetrievals` should be really a function so we do not duplicate code.

Reply:
Yes. Will fix in next version.

Reply:
What next version? You don't want this to be merged? Can we close this PR then?